### DISCUSSION TITLE
Update Middleware Documentation to use Starlette's way of instantiating middleware

### DISCUSSION THREAD

**Discussion by gareth-leake** [fastapi/fastapi#8530] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
app.add_middleware(Middleware1)
app.add_middleware(Middleware2)
```

### Description

I noticed that my middleware was running multiple times, and after some extensive Googleing and reading Starlette and FastAPI docs, I came across [this issue](https://github.com/encode/starlette/discussions/1161) in the Starlette repo.

The general issue is that when you call `add_middleware()` multiple times, it results in re-initializing the middleware each time. While it's not well documented in Starlette, the authors left a comment in the [source code](https://github.com/encode/starlette/blob/1225ae8404ed0ae0b9170fde1cf824ec3f9c0f2f/starlette/applications.py#L126) that `add_middleware()` is not preferred, and instead should be handled during configuration. Presently, FastAPI makes this possible, because you're able to do `FastAP(middleware=[MyCustomMiddleware])`.

However, in the FastAPI documentation, `app.add_middleware()` is what is shown. I think this can cause some issues with middleware usage in FastAPI, and it would help to update the docs and point folks towards the Starlette way of instantiating middleware (i.e. passing a list of middleware to your FastAPI/Starlette instance at initialization).

### Operating System

macOS

### FastAPI Version

0.72.0

### Python Version

3.9.12

**Comment by JarroVGIT** [comment]
I would agree, we should at least make sure the documentation uses the recommended approach.

**Comment by gareth-leake** [comment]
I know FastAPI also has a decorator approach `app.add_middleware("http")` -- I didn't see this approach in the Starlette docs either, so I'm not sure if this is a custom FastAPI approach. But that method _also_ causes multiple instantiations of middleware.

Perhaps this isn't the case if you only call it one time and have one app. But if (1) you have multiple sub-apps mounted onto a base app, and you want middleware in all of them, or (2) you have more than one piece of middleware, both the decorator approach and `app.add_middleware()` approach cause multiple instantiations.

My own solution that worked well was to create a `middleware.py` file, and write all my middleware as Class instances, following the patterns found in Starlette's code.

Example:
```python
# middleware.py

import uuid
import time
import logging
from typing import Callable
from fastapi import Request
from starlette.middleware import Middleware
from starlette.middleware.base import BaseHTTPMiddleware


class RequestIDMiddleware(BaseHTTPMiddleware):
    async def dispatch(self, request: Request, call_next: Callable):
        request.state.uuid = uuid.uuid4().hex
        start_time = time.time()
        response = await call_next(request)
        process_time = time.time() - start_time
        response.headers["X-Process-Time"] = str(process_time)
        logger.debug(f"LATENCY {process_time}s {request.method} {request.url}")
        return response

custom_middleware = [Middleware(RequestIDMiddleware)]
```

```python
#main.py
from .middleware import custom_middleware

app = FastAPI(middleware=custom_middleware)
```

**Comment by JarroVGIT** [comment]
The `app.middleware("http")` decorator is indeed syntax sugar from Starlette (not FastAPI) where the decorated function is treated as the `dispatch` method of a new `BaseHTTPMiddleware` class. It is also discouraged to use in their source code:

See here:
https://github.com/encode/starlette/blob/1225ae8404ed0ae0b9170fde1cf824ec3f9c0f2f/starlette/applications.py#L262-L283

Also, @gareth-leake, be aware that there are discussions going on on deprecating BaseHTTPMiddleware. It's usage is discouraged as well, because it is fundamentally broken. It will not play nicely with contextVars. Recommendation is to rewrite as "pure ASGI" middleware.

**Comment by gareth-leake** [comment]
Ah, yeah that totally makes sense. It's still just calling `add_middleware`. Thanks for pointing that out!

**Comment by Kludex** [comment]
The ~`add_middleware`~ `middleware("http")` decorator is discouraged on Starlette, but not on FastAPI. If removed from Starlette, FastAPI will probably implement the decorators anyway...

The reason for being discouraged is not the "issue" about being initialized multiple times. It's because of the decorator style that Starlette doesn't want to maintain.

**Comment by JarroVGIT** [comment]
> The `add_middleware` is discouraged on Starlette, not on FastAPI.

AFAICT there is no `add_middleware()` method in FastAPI, it inherits it from Starlette directly. I also didn't know that the `Starlette.add_middleware()` itself was discouraged, the source code only discourages the use of the decorator style.

**Comment by Kludex** [comment]
Sorry, I meant `middleware("http")` was discouraged... I've fixed the above. 😅

**Comment by gareth-leake** [comment]
The decorator style ultimately implements the `add_middleware` function, which is discouraged by Starlette [here](https://github.com/encode/starlette/blob/1225ae8404ed0ae0b9170fde1cf824ec3f9c0f2f/starlette/applications.py#L126). Unfortunately the issues around middleware are not well documented. Perhaps this should be something I bring up with Starlette instead. The difficulty is that `add_middleware` and the decorator method both work, _but_ they end up creating issues if called multiple times. However, there's no clear indication in the documentation that these things should only be called once.

**Comment by pattrickrice** [comment]
Just a comment on the experience of using `app.middleware`, this is what the intellisense looks like on the function. 
<img width="907" alt="Screenshot 2023-09-21 at 10 42 20 AM" src="https://github.com/tiangolo/fastapi/assets/20763171/174cd757-49e7-48dd-9299-563d10e5ced1">

**Comment by YuriiMotov** [comment, marked as answer]
So, the problem of re-initializing the middleware each time when you call `add_middleware()` multiple times was solved: https://github.com/encode/starlette/issues/2002

And, use of `add_middleware` is still not discouraged in FastAPI
